### ISSUE TITLE
Disclosure flow on iOS sometimes results in promise rejection

### ISSUE THREAD

**Issue by Timendus** [privacybydesign/irmajs#2]
I ran into this when working on nuts-foundation/irma-web-glue#2. I hope you can reproduce it on your end.

Running the `irmajs` browser example on iOS results in unpredictable behaviour. I see three different possible end states:

# 1. Rejected promise "DONE"

This happens when I kill the IRMA app and go through the disclosure flow.

![IMG_40C98354E27C-1](https://user-images.githubusercontent.com/2649939/56691786-10f9ec80-66e1-11e9-8d6a-5b4acd8e1d6c.jpeg)

### Safari console

```
[Log] Session started – {u: "http://c06098be.ngrok.io/irma/ouuQtsrQDsyZecHg5MQK", irmaqr: "disclosing"} (irma.js, line 9)
[Log] Options: – {server: "http://c06098be.ngrok.io", token: "EtOAB5iY2Du8w8HAyZJD", method: "popup", …} (irma.js, line 9)
{server: "http://c06098be.ngrok.io", token: "EtOAB5iY2Du8w8HAyZJD", method: "popup", language: "en"}Object
[Log] Detected iOS (irma.js, line 9)
[Log] On mobile; using method mobile instead of popup (irma.js, line 9)
[Log] Navigating: – "irma://qr/json/%7B%22u%22%3A%22http%3A%2F%2Fc06098be.ngrok.io%2Firma%2FouuQtsrQDsyZecHg5MQK%22%2C%22irmaqr%22%3A%22disclosing%22%7D" (irma.js, line 9)
[Log] Received server event error – Event {isTrusted: true, type: "error", target: EventSource, …} (irma.js, line 9)
Event {isTrusted: true, type: "error", target: EventSource, currentTarget: EventSource, eventPhase: 2, …}Event
[Log] error in server sent event, falling back to polling: – Event {isTrusted: true, type: "error", target: EventSource, …} (irma.js, line 9)
Event {isTrusted: true, type: "error", target: EventSource, currentTarget: EventSource, eventPhase: 2, …}Event
[Error] Unhandled Promise Rejection: DONE
	(anonymous function)
	promiseReactionJob
```

# 2. Rejected promise with connection error

This happens most of the time when I have my IRMA app already open.

![IMG_2D3C208D3F2F-1](https://user-images.githubusercontent.com/2649939/56691961-61714a00-66e1-11e9-8cc3-122ba077da76.jpeg)

### Safari console

```
[Log] Session started – {u: "http://c06098be.ngrok.io/irma/2iIMsCMdczbnfI5V3tzg", irmaqr: "disclosing"} (irma.js, line 9)
[Log] Options: – {server: "http://c06098be.ngrok.io", token: "6lYEqcr4IUEGaJetQFqy", method: "popup", …} (irma.js, line 9)
{server: "http://c06098be.ngrok.io", token: "6lYEqcr4IUEGaJetQFqy", method: "popup", language: "en"}Object
[Log] Detected iOS (irma.js, line 9)
[Log] On mobile; using method mobile instead of popup (irma.js, line 9)
[Log] Navigating: – "irma://qr/json/%7B%22u%22%3A%22http%3A%2F%2Fc06098be.ngrok.io%2Firma%2F2iIMsCMdczbnfI5V3tzg%22%2C%22irmaqr%22%3A%22disclosing%22%7D" (irma.js, line 9)
[Log] Received server event error – Event {isTrusted: true, type: "error", target: EventSource, …} (irma.js, line 9)
Event {isTrusted: true, type: "error", target: EventSource, currentTarget: EventSource, eventPhase: 2, …}Event
[Log] error in server sent event, falling back to polling: – Event {isTrusted: true, type: "error", target: EventSource, …} (irma.js, line 9)
Event {isTrusted: true, type: "error", target: EventSource, currentTarget: EventSource, eventPhase: 2, …}Event
[Error] Failed to load resource: De bewerking kan niet worden voltooid. Verbinding verbroken door software
[Error] Unhandled Promise Rejection: TypeError: De bewerking kan niet worden voltooid. Verbinding verbroken door software
	(anonymous function)
	promiseReactionJob
```

# 3. Resolved promise (as intended)

This happens sometimes when I have the IRMA app already open.

![IMG_43C3552B7C60-1](https://user-images.githubusercontent.com/2649939/56692259-173c9880-66e2-11e9-94d7-e320412bf3b2.jpeg)

### Safari console

```
[Log] Session started – {u: "http://c06098be.ngrok.io/irma/ar4BibBrDpw8qzvahrvF", irmaqr: "disclosing"} (irma.js, line 9)
[Log] Options: – {server: "http://c06098be.ngrok.io", token: "cKpaJyf6r6Qi3tKTkhbU", method: "popup", …} (irma.js, line 9)
{server: "http://c06098be.ngrok.io", token: "cKpaJyf6r6Qi3tKTkhbU", method: "popup", language: "en"}Object
[Log] Detected iOS (irma.js, line 9)
[Log] On mobile; using method mobile instead of popup (irma.js, line 9)
[Log] Navigating: – "irma://qr/json/%7B%22u%22%3A%22http%3A%2F%2Fc06098be.ngrok.io%2Firma%2Far4BibBrDpw8qzvahrvF%22%2C%22irmaqr%22%3A%22disclosing%22%7D" (irma.js, line 9)
[Log] Received server event error – Event {isTrusted: true, type: "error", target: EventSource, …} (irma.js, line 9)
Event {isTrusted: true, type: "error", target: EventSource, currentTarget: EventSource, eventPhase: 2, …}Event
[Log] error in server sent event, falling back to polling: – Event {isTrusted: true, type: "error", target: EventSource, …} (irma.js, line 9)
Event {isTrusted: true, type: "error", target: EventSource, currentTarget: EventSource, eventPhase: 2, …}Event
[Log] Session state changed – "CONNECTED" – "http://c06098be.ngrok.io/irma/ar4BibBrDpw8qzvahrvF" (irma.js, line 9)
[Log] error in server sent event, falling back to polling: – "no open message received" (irma.js, line 9)
[Log] Received server event error – Event {isTrusted: true, type: "error", target: EventSource, …} (irma.js, line 9)
Event {isTrusted: true, type: "error", target: EventSource, currentTarget: EventSource, eventPhase: 2, …}Event
[Log] Done – {token: "cKpaJyf6r6Qi3tKTkhbU", status: "DONE", type: "disclosing", …} (16f00669.ngrok.io, line 51)
{token: "cKpaJyf6r6Qi3tKTkhbU", status: "DONE", type: "disclosing", proofStatus: "VALID", disclosed: Array}Object
```

**Comment by Timendus** [comment]
Interestingly enough this does not seem to happen in an installed Progressive Web App on iOS. The app gets reloaded fully and gives the proper result.

**Comment by davidv1992** [comment]
Sorry for the late response, I will look into it later this week.

**Comment by davidv1992** [comment]
Took some time to find the cause, but we understand what goes wrong now, fix should come later today or early next week

**Comment by Timendus** [comment]
Ooh, that's awesome news! 😄

**Comment by sietseringers** [comment]
The problem was that if the browser is backgrounded and paused, then the session would be finished instead of halfway when control is returned to the browser, taking the polling function by surprise. This should now be fixed in `v0.1.8`.
